File: client/src/app/_models/subscriptions/subscription.ts
import { Entity } from "src/app/_models/base/entity";
import { SubscriptionHistory } from "src/app/_models/subscriptionHistories/subscriptionHistory";
import { SubscriptionStatus } from "src/app/_models/subscriptions/subscriptionConstants";

export class Subscription extends Entity {
  userId!: number;
  planId!: number;
  price!: number;
  planName!: string;
  startDate!: Date;
  endDate?: Date;
  status!: SubscriptionStatus;
  nextBillingDate?: Date;
  stripeSubscriptionId?: string;
  stripeCustomerId?: string;
  stripePlanId?: string;
  histories?: SubscriptionHistory[];

  constructor(init?: Partial<Subscription>) {
    super();
    Object.assign(this, init);
  }
}

File: src/MainService.Models/Entities/Subscription.cs
namespace MainService.Models.Entities;

public class Subscription : BaseEntity
{
    public int UserId { get; set; }
    public AppUser User { get; set; } = null!;

    public int SubscriptionPlanId { get; set; }
    public SubscriptionPlan SubscriptionPlan { get; set; } = null!;

    public DateTime SubscriptionStartDate { get; set; }
    public DateTime? SubscriptionEndDate { get; set; }
    public SubscriptionStatus Status { get; set; }
    public DateTime? NextBillingDate { get; set; }

    public string? StripeSubscriptionId { get; set; }
    public string? StripeCustomerId { get; set; }

    public List<SubscriptionHistory> SubscriptionHistories { get; set; } = [];
}